### PR TITLE
Add methods for getting entity sounds for an entity

### PR DESCRIPTION
**Pull request by Maxqia**
**API** | [Common](https://github.com/SpongePowered/SpongeCommon/pull/1026)
I had this lying around for a month or so..., so here you go!
There might be some sounds missing and if there are, please tell me where I can find them.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/entity/Entity.java [hunk 1]
```diff
@@ -560,4 +561,17 @@ default boolean canSee(Entity entity) {
...
+     */
+    SoundType getSwimSound();
+
+    /**
+     * Gets the {@link SoundType} that is played when this entity jumps in water.
+     *
+     * @return The SoundType for the noise this entity makes when it enters water
+     */
+    SoundType getSplashSound();
 }
```
Comment (liach) [inline src/main/java/org/spongepowered/api/entity/Entity.java:577]: You need an extra line before this line.

File: src/main/java/org/spongepowered/api/entity/living/Living.java [hunk 2]
```diff
@@ -110,6 +111,43 @@ default DamageableData getMortalData() {
...
+     * Gets the {@link SoundType} that is played when this entity dies.
+     *
+     * @return The {@link SoundType} that is played when this entity dies
+     */
+    SoundType getDeathSound();
+
+    /**
+     * Gets the {@link SoundType} that is played when this entity falls
+     * This return a big or small fall sound based on how far the entity is specified to fall.
+     * @param distance How far the entity fell
```
Comment (liach) [inline src/main/java/org/spongepowered/api/entity/living/Living.java:145]: You need an extra empty line of Javadoc before this line.

File: src/main/java/org/spongepowered/api/entity/living/Living.java [hunk 3]
```diff
@@ -110,6 +111,43 @@ default DamageableData getMortalData() {
...
+     * @return The {@link SoundType} that is played when this entity falls that distance
+     */
+    SoundType getFallSound(int distance);
```
Comment (liach) [inline src/main/java/org/spongepowered/api/entity/living/Living.java:148]: You should probably use a `double` for such distance.

**Review by liach — CHANGES_REQUESTED** [review]

**Comment by liach** [comment]
In general, the best way to retrieve a sound is to provide a context.

**Comment by ST-DDT** [comment]
In https://github.com/SpongePowered/SpongeAPI/pull/1376 / https://github.com/SpongePowered/SpongeAPI/commit/4ae202fa5cfb99824177851c45352863640d3704 we went for `BlockSoundGroup` maybe we should do the same here for `EntitySoundGroup`?

Maybe we could also structure it a little bit:

soundgroup
[Picture/DOT-SourceCode](http://pastebin.com/keUkcCCe)

How about screaming/growling/angry pigmen...
Also there is the "idle" sound. Afaict its called `getAmbientSound()`

[Here](http://pastebin.com/LqrDygjD) is a list of remaining mob sounds for 1.10 (without death, hit and hurt)

**Comment by ST-DDT** [comment]
> In general, the best way to retrieve a sound is to provide a context.

@liach What context are you referring to?

File: src/main/java/org/spongepowered/api/entity/living/Living.java [hunk 4]
```diff
@@ -110,6 +111,43 @@ default DamageableData getMortalData() {
     }
 
     /**
+     * Gets the volume that the sounds for this entity are played
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/living/Living.java:114]: Missing dot at the end of sentence.

File: src/main/java/org/spongepowered/api/entity/living/Living.java [hunk 5]
```diff
@@ -110,6 +111,43 @@ default DamageableData getMortalData() {
...
+     *
+     * @return The volume of the sounds for this entity
+     */
+    double getVolume();
+
+    /**
+     * Gets the pitch for the sounds for this entity
+     * (e.g for if it's a baby)
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/living/Living.java:122]: Missing dot at the end of sentence.

File: src/main/java/org/spongepowered/api/entity/living/Living.java [hunk 6]
```diff
@@ -110,6 +111,43 @@ default DamageableData getMortalData() {
...
+     *
+     * @return The pitch of the sounds for this entity
+     */
+    double getPitch();
+
+    /**
+     * Gets the {@link SoundType} that is played when this entity gets hurt.
+     *
+     * @return The {@link SoundType} that is played when this entity gets hurt
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/living/Living.java:131]: I guess its not necessary to link to the returned class here again.

File: src/main/java/org/spongepowered/api/entity/living/Living.java [hunk 7]
```diff
@@ -110,6 +111,43 @@ default DamageableData getMortalData() {
...
+     */
+    SoundType getHurtSound();
+
+    /**
+     * Gets the {@link SoundType} that is played when this entity dies.
+     *
+     * @return The {@link SoundType} that is played when this entity dies
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/living/Living.java:138]: I guess its not necessary to link to the returned class here again.

File: src/main/java/org/spongepowered/api/entity/living/Living.java [hunk 8]
```diff
@@ -110,6 +111,43 @@ default DamageableData getMortalData() {
...
+     */
+    SoundType getDeathSound();
+
+    /**
+     * Gets the {@link SoundType} that is played when this entity falls
+     * This return a big or small fall sound based on how far the entity is specified to fall.
+     * @param distance How far the entity fell
+     * @return The {@link SoundType} that is played when this entity falls that distance
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/living/Living.java:146]: I guess its not necessary to link to the returned class here again.

File: src/main/java/org/spongepowered/api/entity/living/Living.java (hunk 3 above)
Comment (ST-DDT) [reply src/main/java/org/spongepowered/api/entity/living/Living.java:148]: Just curious: Is there a difference of sound depending on the fall distance? (I haven't noticed it myself yet, but I just may be deaf.^^) Maybe we can expose the short and long sounds directly.

File: src/main/java/org/spongepowered/api/entity/Entity.java [hunk 9]
```diff
@@ -560,4 +561,17 @@ default boolean canSee(Entity entity) {
         return getValue(Keys.HAS_GRAVITY).get();
     }
 
+    /**
+     * Gets the {@link SoundType} that is played when this entity moves in water.
+     *
+     * @return The SoundType for noise this entity makes when it swims
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/Entity.java:567]: IMO these sentences could be simplified to:

```
The sound that the entity emits when it ...
```

Same goes for the other methods. Not 100% sure though.

**Review by ST-DDT — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/api/entity/living/Living.java (hunk 3 above)
Comment (liach) [reply src/main/java/org/spongepowered/api/entity/living/Living.java:148]: Yes, if you have noticed vanilla sound types. There is a small fall and a big fall.

**Comment by Maxqia** [comment]
I did it mainly for simplicity of the implementation, because the methods and fields for getting sounds for entities are in the entity class, while block sounds are in a different class.

**Comment by stephan-gh** [comment]
@Maxqia 
In my opinion the structure @ST-DDT suggested makes sense and is actually quite simple to implement:

It shouldn't take long to create another class in the implementation that links to the internal Entity class and just returns the sounds. If you really have to you could even make the internal Entity class implement `EntitySoundGroup` directly and just return `this` when `getSoundGroup()` is called on `Entity`.

The purpose of SpongeAPI is not to represent the internal classes exactly but to hide most of the internal structure and provide it in a logical way to plugin developers.

**Comment by Maxqia** [comment]
I might make it actually so that BlockSoundGroup is integrated into BlockType, because making a sound group for (practically) every entity is going to make me lose my sanity :P
EDIT : it also makes a little more sense logically

**Comment by kashike** [comment]
@Maxqia Status on the comments above?

**Comment by kashike** [comment]
@Maxqia Status on the comments above?

**Comment by Zidane** [comment]
@kashike is taking over this PR.

**Pull request closed without merging**